pkg/network/usm/istio.go

```diff
@@ -0,0 +1,221 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package usm
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/network/usm/utils"
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	manager "github.com/DataDog/ebpf-manager"
+)
+
+// envoyCmd represents the search term used for determining
+// whether or not a given PID represents an Envoy process.
+// The search is done over the /proc/<pid>/cmdline file.
+var envoyCmd = []byte("/bin/envoy")
+
+// readBufferPool is used for reading /proc/<pid>/cmdline files.
+// We use a pointer to a slice to avoid allocations when casting
+// values to the empty interface during Put() calls.
+var readBufferPool = sync.Pool{
+	New: func() any {
+		b := make([]byte, 128)
+		return &b
+	},
+}
+
+// istioMonitor essentially scans for Envoy processes and attaches SSL uprobes
+// to them.
+//
+// Note that for now we only support Istio as opposed to "vanilla" Envoy
+// because the Envoy binary embedded in the Istio containers have debug symbols
+// whereas the "vanilla" Envoy images are distributed without them.
+type istioMonitor struct {
+	registry *utils.FileRegistry
+	procRoot string
+
+	// `utils.FileRegistry` callbacks
+	registerCB   func(utils.FilePath) error
+	unregisterCB func(utils.FilePath) error
+
+	// Termination
+	wg   sync.WaitGroup
+	done chan struct{}
+}
+
+func newIstioMonitor(c *config.Config, mgr *manager.Manager) *istioMonitor {
+	if !c.EnableIstioMonitoring {
+		return nil
+	}
+
+	return &istioMonitor{
+		registry: utils.NewFileRegistry("istio"),
+		procRoot: util.GetProcRoot(),
+		done:     make(chan struct{}),
+
+		// Callbacks
+		registerCB:   addHooks(mgr, openSSLProbes),
+		unregisterCB: removeHooks(mgr, openSSLProbes),
+	}
+}
+
+func (m *istioMonitor) Start() {
+	if m == nil {
+		return
+	}
+
+	processMonitor := monitor.GetProcessMonitor()
+
+	// Subscribe to process events
+	processMonitor.SubscribeExec(m.handleProcessExec)
+	processMonitor.SubscribeExit(m.handleProcessExit)
```

Comment:
this is a bug, you are not saving the cleanup functions for the callbacks, and you must unregister them before terminating the loop.

Reply:
oof good catch 👍